File: messaging-common/src/main/java/com/hmdev/messaging/common/security/MySecurity.java
package com.hmdev.messaging.common.security;

import com.hmdev.messaging.common.crypto.EnvelopeUtil;
import com.hmdev.messaging.common.security.aes.AesCtr;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.KeySpec;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

public final class MySecurity {
    private static final Logger logger = LoggerFactory.getLogger(MySecurity.class);

    private static final byte[] PBKDF2_SALT = "messaging-platform".getBytes(StandardCharsets.UTF_8);
    private static final int PBKDF2_ITERATIONS = 100_000;
    private static final int PBKDF2_KEY_LENGTH = 256; // bits
    private static final SecretKeyFactory PBKDF2_FACTORY;
    private static final Base64.Encoder B64_URL_NO_PADDING = Base64.getUrlEncoder().withoutPadding();

    // Performance-oriented shared resources
    private static final Base64.Encoder B64 = Base64.getEncoder();
    private static final ThreadLocal<Mac> MAC_THREAD_LOCAL = ThreadLocal.withInitial(() -> {
        try {
            return Mac.getInstance("HmacSHA256");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    });
    private static final char[] HEX_ARRAY = "0123456789abcdef".toCharArray();

    // Small cache to avoid recreating SecretKeySpec objects for the same key strings
    private static final ConcurrentHashMap<String, SecretKeySpec> KEY_SPEC_CACHE = new ConcurrentHashMap<>();

    static {
        try {
            PBKDF2_FACTORY = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
        } catch (NoSuchAlgorithmException exception) {
            throw new RuntimeException(exception);
        }
    }

    private MySecurity() { /* no instances */ }

    public static String encryptAndSign(String message, String key) {
        JSONObject $myObj = new JSONObject("{}");
        $myObj.put("cipher", encrypt(message, key));
        $myObj.put("hash", hash(message, key));
        return $myObj.toString();
    }

    public static String decryptAndVerify(String cipherMsgStr, String key) {
        try {
            JSONObject cipherMsg = new JSONObject(cipherMsgStr);
            String message = decrypt(cipherMsg.optString("cipher"), key);
            if (!Objects.equals(hash(message, key), cipherMsg.optString("hash"))) {
                return null;
            } else {
                return message;
            }
        } catch (Exception e) {
            logger.debug("decryptWithMd5Auth error: {}", e.getMessage());
            return null;
        }
    }

    public static String encrypt(String $plain, String $key) {
        try {
            return AesCtr.encrypt($plain, $key, 128);
        } catch (Exception e) {
            logger.debug("encrypt error: {}", e.getMessage());
            return "";
        }
    }

    public static String decrypt(String $cipher, String $key) {
        try {
            return AesCtr.decrypt($cipher, $key, 128);
        } catch (Exception e) {
            logger.debug("decrypt error: {}", e.getMessage());
        }
        return null;
    }

    public static String deriveChannelSecret(String channelName, String password) {
        String combined = channelName + password;
        KeySpec spec = new PBEKeySpec(combined.toCharArray(), PBKDF2_SALT, PBKDF2_ITERATIONS, PBKDF2_KEY_LENGTH);
        try {
            byte[] keyBytes = PBKDF2_FACTORY.generateSecret(spec).getEncoded();
            return "channel_" + B64_URL_NO_PADDING.encodeToString(keyBytes);
        } catch (InvalidKeySpecException e) {
            throw new RuntimeException("Key derivation failed", e);
        }
    }

    // ------------------ RSA helpers (delegates to EnvelopeUtil) ------------------
    public static KeyPair rsaGenerate() throws Exception {
        return EnvelopeUtil.generateRSAKeyPair();
    }

    public static String rsaEncodePublicKey(PublicKey pub) {
        return EnvelopeUtil.encodeRSAPublicKey(pub);
    }

    public static byte[] rsaEncrypt(PublicKey pub, byte[] plaintext) throws Exception {
        return EnvelopeUtil.rsaEncrypt(pub, plaintext);
    }

    public static byte[] rsaDecrypt(PrivateKey priv, byte[] ciphertext) throws Exception {
        return EnvelopeUtil.rsaDecrypt(priv, ciphertext);
    }

    // New overloads: accept keys as objects or strings and plaintext/ciphertext as raw Strings
    public static String rsaEncrypt(PublicKey pub, String plaintext) throws Exception {
        byte[] ct = rsaEncrypt(pub, plaintext.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(ct);
    }

    public static String rsaEncrypt(String pubKeyStr, String plaintext) throws Exception {
        PublicKey pub = publicKeyFromString(pubKeyStr);
        return rsaEncrypt(pub, plaintext);
    }

    public static String rsaDecrypt(PrivateKey priv, String cipherB64) throws Exception {
        byte[] cipher = Base64.getDecoder().decode(cipherB64);
        byte[] plain = rsaDecrypt(priv, cipher);
        return new String(plain, StandardCharsets.UTF_8);
    }

    public static String rsaDecrypt(String privKeyStr, String cipherB64) throws Exception {
        PrivateKey priv = privateKeyFromString(privKeyStr);
        return rsaDecrypt(priv, cipherB64);
    }

    private static PublicKey publicKeyFromString(String keyStr) throws Exception {
        String s = keyStr.trim();
        byte[] keyBytes;
        if (s.contains("-----BEGIN")) {
            // PEM format: strip header/footer
            s = s.replaceAll("-----BEGIN (?:.*)-----", "");
            s = s.replaceAll("-----END (?:.*)-----", "");
            s = s.replaceAll("\\s", "");
            keyBytes = Base64.getDecoder().decode(s);
        } else {
            try {
                keyBytes = Base64.getUrlDecoder().decode(s);
            } catch (IllegalArgumentException e) {
                keyBytes = Base64.getDecoder().decode(s);
            }
        }
        KeyFactory kf = KeyFactory.getInstance("RSA");
        return kf.generatePublic(new X509EncodedKeySpec(keyBytes));
    }

    private static PrivateKey privateKeyFromString(String keyStr) throws Exception {
        String s = keyStr.trim();
        byte[] keyBytes;
        if (s.contains("-----BEGIN")) {
            // PEM format: strip header/footer
            s = s.replaceAll("-----BEGIN (?:.*)-----", "");
            s = s.replaceAll("-----END (?:.*)-----", "");
            s = s.replaceAll("\\s", "");
            keyBytes = Base64.getDecoder().decode(s);
        } else {
            try {
                keyBytes = Base64.getUrlDecoder().decode(s);
            } catch (IllegalArgumentException e) {
                keyBytes = Base64.getDecoder().decode(s);
            }
        }
        KeyFactory kf = KeyFactory.getInstance("RSA");
        return kf.generatePrivate(new PKCS8EncodedKeySpec(keyBytes));
    }

    public static String blocksEncrypt(Cipher encryptor, String plain) {
        if (encryptor == null) {
            return plain;
        }
        try {
            // Work on UTF-8 bytes once to avoid repeated substring allocations
            byte[] plainBytes = plain.getBytes(StandardCharsets.UTF_8);
            StringBuilder cipherPayload = new StringBuilder((plainBytes.length/200 + 1) * 24);
            for (int i = 0; i < plainBytes.length; i += 200) {
                int upperLimit = Math.min(i + 200, plainBytes.length);
                int len = upperLimit - i;
                // avoid allocating a temporary chunk array by using the doFinal overload
                byte[] salted = encryptor.doFinal(plainBytes, i, len);
                cipherPayload.append(B64.encodeToString(salted));
            }
            return cipherPayload.toString();
        } catch (Exception e) {
            logger.warn(e.getMessage());
            return "";
        }
    }

    public static String hash(String msg, String key) {
        if (msg == null || key == null) {
            return null;
        }
        try {
            Mac mac = MAC_THREAD_LOCAL.get();
            // cache SecretKeySpec to avoid recreating the object for frequently used keys
            SecretKeySpec secretKeySpec = KEY_SPEC_CACHE.computeIfAbsent(key, k ->
                    new SecretKeySpec(k.getBytes(StandardCharsets.UTF_8), "HmacSHA256")
            );
            mac.init(secretKeySpec);
            byte[] hmacBytes = mac.doFinal(msg.getBytes(StandardCharsets.UTF_8));
            return toHexString(hmacBytes); // keep your existing hex converter
        } catch (Exception e) {
            logger.warn(e.getMessage());
            return null;
        }
    }

    public static String toHexString(byte[] bytes) {
        // faster hex conversion, fewer allocations than StringBuilder+Integer.toHexString
        char[] hexChars = new char[bytes.length * 2];
        for (int j = 0; j < bytes.length; j++) {
            int v = bytes[j] & 0xFF;
            hexChars[j * 2] = HEX_ARRAY[v >>> 4];
            hexChars[j * 2 + 1] = HEX_ARRAY[v & 0x0F];
        }
        return new String(hexChars);
    }
}

File: agents/python-agent/hmdev/messaging/agent/security/my_security.py
import base64
import hmac, hashlib
import json
import logging

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import serialization

from hmdev.messaging.agent.security.aes_ctr import AesCtr

logger = logging.getLogger(__name__)


class MySecurity:
    """Static-like utility class for encryption, signing, and verification."""

    def __init__(self):
        raise RuntimeError("No instances allowed")

    @staticmethod
    def encrypt_and_sign(message: str, key: str) -> str:
        data = {
            "cipher": MySecurity.encrypt(message, key),
            "hash": MySecurity.hash(message, key),
        }
        return json.dumps(data)

    @staticmethod
    def decrypt_and_verify(cipher_msg_str: str, key: str) -> str | None:
        try:
            cipher_msg = json.loads(cipher_msg_str)
            message = MySecurity.decrypt(cipher_msg.get("cipher", ""), key)

            if message is None:
                return None

            if MySecurity.hash(message, key).strip() != cipher_msg.get("hash", "").strip():
                return None
            return message
        except Exception as e:
            logger.debug(f"decrypt_with_md5_auth error: {e}")
            return None

    # ------------------ AES CTR helper ------------------

    @staticmethod
    def encrypt(plain: str, key: str) -> str:
        try:
            return AesCtr.encrypt(plain, key, 128)
        except Exception as e:
            logger.debug(f"encrypt error: {e}")
            return ""

    @staticmethod
    def decrypt(cipher: str, key: str) -> str | None:
        try:
            return AesCtr.decrypt(cipher, key, 128)
        except Exception as e:
            logger.debug(f"decrypt error: {e}")
            return None

    # ------------------ RSA helpers ------------------
    @staticmethod
    def rsa_generate(key_size: int = 2048):
        """Generate RSA private key and return (private_key_obj, public_key_pem_str)."""
        priv = rsa.generate_private_key(public_exponent=65537, key_size=key_size)
        pub_pem = priv.public_key().public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        ).decode('utf-8')
        return priv, pub_pem

    @staticmethod
    def _load_public_key(key_str: str):
        s = key_str.strip()
        try:
            if "-----BEGIN" in s:
                # PEM
                return serialization.load_pem_public_key(s.encode('utf-8'), backend=default_backend())
            else:
                # try URL-safe base64 first, then regular base64
                try:
                    data = base64.urlsafe_b64decode(s + '==')
                except Exception:
                    data = base64.b64decode(s)
                return serialization.load_der_public_key(data, backend=default_backend())
        except Exception as e:
            logger.debug(f"_load_public_key error: {e}")
            raise

    @staticmethod
    def _load_private_key(key_str: str):
        s = key_str.strip()
        try:
            if "-----BEGIN" in s:
                # PEM
                return serialization.load_pem_private_key(s.encode('utf-8'), password=None, backend=default_backend())
            else:
                # try URL-safe base64 then regular base64
                try:
                    data = base64.urlsafe_b64decode(s + '==')
                except Exception:
                    data = base64.b64decode(s)
                return serialization.load_der_private_key(data, password=None, backend=default_backend())
        except Exception as e:
            logger.debug(f"_load_private_key error: {e}")
            raise

    @staticmethod
    def rsa_encrypt(public_key, plaintext: str) -> str:
        """Encrypt plaintext string with a public key.

        public_key may be either a PEM/base64 string or a public key object.
        Returns base64 ciphertext string.
        """
        try:
            if isinstance(public_key, str):
                pub = MySecurity._load_public_key(public_key)
            else:
                pub = public_key

            ct = pub.encrypt(
                plaintext.encode('utf-8'),
                padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None)
            )
            return base64.b64encode(ct).decode('ascii')
        except Exception as e:
            logger.debug(f"rsa_encrypt error: {e}")
            raise

    @staticmethod
    def rsa_decrypt(private_key, base64_cipher: str) -> str:
        """Decrypt base64 ciphertext with a private key.

        private_key may be either a PEM/base64 string or a private key object. Returns UTF-8 plaintext.
        """
        try:
            if isinstance(private_key, str):
                priv = MySecurity._load_private_key(private_key)
            else:
                priv = private_key

            ct = base64.b64decode(base64_cipher)
            plain = priv.decrypt(
                ct,
                padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None)
            )
            return plain.decode('utf-8')
        except Exception as e:
            logger.debug(f"rsa_decrypt error: {e}")
            raise

    # ------------------ Key derivation ------------------

    @staticmethod
    def derive_channel_secret(channel_name: str, password: str) -> str:
        combined = (channel_name + password).encode()
        salt = b"messaging-platform"  # must match JS
        iterations = 100_000
        key_length = 32  # 256 bits

        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=key_length,
            salt=salt,
            iterations=iterations,
            backend=default_backend(),
        )
        key_bytes = kdf.derive(combined)
        return 'channel_' + base64.urlsafe_b64encode(key_bytes).rstrip(b'=').decode('ascii')

    # ------------------ MD5 Hash ------------------

    @staticmethod
    def hash(msg: str, key_str: str) -> str:
        try:
            # return hashlib.sha256(msg.encode("utf-8")).hexdigest()
            return hmac.new(key_str.encode("utf-8"), msg.encode("utf-8"), hashlib.sha256).hexdigest()
        except Exception as e:
            logger.warning(f"hash error: {e}")
            return None
